blocks/library/reusable-block/index.js

```diff
@@ -0,0 +1,154 @@
+
+		this.state = {
+			isEditing: false,
+			name: null,
+			attributes: null,
+		};
+	}
+
+	componentDidMount() {
+		if ( ! this.props.reusableBlock ) {
+			this.props.fetchReusableBlock();
+		}
+	}
+
+	toggleEditing() {
+		this.setState( ( prevState ) => ( {
+			isEditing: ! prevState.isEditing,
+			name: null,
+			attributes: null,
+		} ) );
+	}
+
+	setAttributes( attributes ) {
+		this.setState( ( prevState ) => ( {
+			attributes: { ...prevState.attributes, ...attributes },
+		} ) );
+	}
+
+	setName( name ) {
+		this.setState( { name } );
+	}
+
+	updateReusableBlock() {
+		const { name, attributes } = this.state;
+		// TODO: Can these two actions be combined?
+		// TODO: Think about a loading indicator, success message, failure message, etc.
+		this.props.updateReusableBlock( pickBy( { name, attributes } ) );
+		this.props.saveReusableBlock();
+		this.toggleEditing();
+	}
+
+	render() {
+		const { focus, reusableBlock, attachBlock } = this.props;
+		const { isEditing, name, attributes } = this.state;
+
+		if ( ! reusableBlock ) {
+			return <Placeholder><Spinner /></Placeholder>;
+		}
+
+		const blockType = getBlockType( reusableBlock.type );
+		const BlockEdit = blockType.edit || blockType.create;
+
+		return [
+			// We fake the block being read-only by wrapping it with an element that has pointer-events: none
+			<div key="edit" style={ { pointerEvents: isEditing ? 'auto' : 'none' } }>
+				<BlockEdit
+					{ ...this.props }
+					focus={ isEditing ? focus : null }
+					attributes={ { ...reusableBlock.attributes, ...attributes } }
+					setAttributes={ isEditing ? this.setAttributes : noop } />
+			</div>,
+			focus && (
+				<ReusableBlockEditPanel
+					key="panel"
+					isEditing={ isEditing }
+					name={ name !== null ? name : reusableBlock.name }
+					onEdit={ this.toggleEditing }
+					onAttach={ attachBlock }
+					onChangeName={ this.setName }
+					onSave={ this.updateReusableBlock }
+					onCancel={ this.toggleEditing } />
+			),
+		];
+	}
+}
+
+/**
+ * TODO:
+ * These should use selectors and action creators that we gather from context.
+ * OR this entire component should move to `editor`, I can't decide.
+ */
+const ConnectedReusableBlockEdit = connect(
+	( state, ownProps ) => ( {
+		reusableBlock: state.reusableBlocks[ ownProps.attributes.ref ],
+	} ),
+	( dispatch, ownProps ) => ( {
+		fetchReusableBlock() {
+			dispatch( {
+				type: 'FETCH_REUSABLE_BLOCK',
+				id: ownProps.attributes.ref,
+			} );
+		},
+		updateReusableBlock( reusableBlock ) {
+			dispatch( {
+				type: 'UPDATE_REUSABLE_BLOCK',
+				id: ownProps.attributes.ref,
+				reusableBlock,
+			} );
+		},
+		saveReusableBlock() {
+			dispatch( {
+				type: 'SAVE_REUSABLE_BLOCK',
+				id: ownProps.attributes.ref,
+			} );
+		},
+		attachBlock() {
+			dispatch( {
+				type: 'ATTACH_BLOCK',
+				uid: ownProps.id,
+			} );
+		},
+	} )
+)( ReusableBlockEdit );
+
+registerBlockType( 'core/reusable-block', {
+	title: __( 'Reusable Block' ),
+	category: null,
```

Comment:
What does a `null` category mean? If I understand correctly, this hides the block from the inserter. this is probably just a side effect of how the inserter is built. Should we be more explicit and add a "private" flag instead?

Reply:
Yeah, that's right. I made it so that `category` can be marked `null` so as to have it not appear in the inserter. 

A more explicit flag is a good idea. I've introduced this in ddc491b32e7e8bcc269caf220401f7005713f9ad.